### PR TITLE
Diagnostics: Adds Processor count into CosmosDiagnostics

### LINKED ISSUE #2978: Diagnostics and Telemetry: Add processor count
The current Diagnostics and Telemetry does not have the processor count, and this is needed because it can be the cause of high latency. The recommendation is to have 4 or more cores, but there is no easy way to confirm the machine has it.

CosmosDiagnostics: recommend adding "ProcessorCount" to the Client Configuration.
```json
"Client Configuration": {
            "Client Created Time Utc": "2022-01-10T18:08:56.7869301Z",
            "NumberOfClientsCreated": 1,
            "User Agent": "cosmos-netstandard-sdk/3.22.0|3.22.0|1|X64|Linux 4.15.0-158-generic 166-|.NET Core 3.1.21|N|",
            "ConnectionConfig": {
                "gw": "(cps:50, urto:10, p:False, httpf: False)",
                "rntbd": "(cto: 5, icto: -1, mrpc: 30, mcpe: 65535, erd: True, pr: ReuseUnicastPort)",
                "other": "(ed:False, be:False)"
            },
            "ConsistencyConfig": "(consistency: NotSet, prgns:[])",
            "ProcessorCount": 4
        }

```

### PR DESCRIPTION
**Pull request by imanvt**
# Pull Request Template

## Description

Adds Processor count into Client configuration section of diagnostics. Insufficient number of processors can cause high latency, so this will help identify those situations.

## Type of change
- [] New feature (non-breaking change which adds functionality)

## Closing issues
closes #2978

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs [hunk 1]
```diff
@@ -105,6 +108,9 @@ private ReadOnlyMemory<byte> GetSerializedDatum()
 
             jsonTextWriter.WriteFieldName("ConsistencyConfig");
             jsonTextWriter.WriteStringValue(this.ConsistencyConfig.ToString());
+            jsonTextWriter.WriteFieldName("ProcessorCount");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:111]: Can you add some tests that verify that this is included in the Diagnostics? There should be such tests that are validating other parts of the Diagnostics. That way, it prevents a future regression if someone accidentally removes this data.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs [hunk 2]
```diff
@@ -104,6 +104,7 @@ public async Task CachedSerializationTest()
             trace = ((CosmosTraceDiagnostics)response.Diagnostics).Value;
             clientConfigurationTraceDatum = (ClientConfigurationTraceDatum)trace.Data["Client Configuration"];
             Assert.IsNotNull(clientConfigurationTraceDatum.SerializedJson);
+            Assert.IsTrue(clientConfigurationTraceDatum.ProcessorCount >= 0);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs:107]: Can it be zero? Or should it be zero? WHy not `>0`?

Comment (imanvt) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs:107]: It looked like these tests were failing when I had it at != 0:
https://cosmos-db-sdk-public.visualstudio.com/cosmos-db-sdk-public/_build/results?buildId=41542&view=logs&j=fb683405-d979-52da-6de9-2541dff429a6&t=0dcb24b2-c512-5a84-18ed-9e0ca54c8d53

Not sure if this is expected behavior or a transient set of test failures.

Comment (imanvt) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs:107]: same tests passed now.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs:107]: The test passed on that build: 
image

THe failure was just environment.

Can you try modifying the test to be > 0 and see if it passes?

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientConfigurationDiagnosticTest.cs:107]: Why not just compare it to Environment processor count value?

**Review by j82w — APPROVED** [review]

**Pull request merged**